server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java

```diff
@@ -287,19 +292,25 @@ public Projector visitOrderedTopN(OrderedTopNProjection projection, Context cont
     }
 
     @Override
-    public Projector visitTopNDistinct(TopNDistinctProjection topNDistinct, Context context) {
+    public Projector visitLimitDistinct(LimitDistinctProjection limitDistinct, Context context) {
         var rowAccounting = new RowCellsAccountingWithEstimators(
-            Symbols.typeView(topNDistinct.outputs()),
+            Symbols.typeView(limitDistinct.outputs()),
             context.ramAccounting,
             0
         );
-        return new TopNDistinctProjector(topNDistinct.limit(), rowAccounting);
+        return new LimitDistinctProjector(limitDistinct.limit(), rowAccounting);
     }
 
     @Override
-    public Projector visitTopNProjection(TopNProjection projection, Context context) {
-        assert projection.limit() > TopN.NO_LIMIT : "TopNProjection must have a limit";
-        return new SimpleTopNProjector(projection.limit(), projection.offset());
+    public Projector visitLimitAndOffsetProjection(LimitAndOffsetProjection projection, Context context) {
+        if (projection.limit() != NO_LIMIT) {
+            return new SimpleLimitAndOffsetProjector(projection.limit(), projection.offset());
+        } else if (projection.offset() != NO_OFFSET) {
+            return new SimpleOffsetProjector(projection.offset());
```

Comment:
Could also consider to have a `LimitProjector` and use both if there is limit-and-offset

Reply:
with the current infrastructure Is there a nice way to do, and not resort to:
```
@Override
    public Projector visitLimitAndOffsetProjection(LimitAndOffsetProjection projection, Context context) {
        final Projector[] projector = new Projector[1];
        if (projection.limit() != NO_LIMIT) {
            projector[0] = new LimitProjector(projection.limit());
        }
        if (projection.offset() != NO_OFFSET) {
            if (projector[0] == null) {
                projector[0] = new OffsetProjector(projection.offset());
            } else {
                projector[0] = rowBatchIterator -> new OffsetProjector(projection.offset()).apply(projector[0].apply(
                    rowBatchIterator));
            }
        }
        if (projector[0] == null) {
            throw new IllegalArgumentException("LimitAndOffsetProjection should have at least one " +
                                               "of the limit and offset set");
        }
        return projector[0];
    }
```